src/core/sock/sock-redirect.cpp

```diff
@@ -306,115 +309,154 @@ bool handle_close(int fd, bool cleanup, bool passthrough)
 
-int init_child_process_for_nginx()
+static int init_worker(int worker_id, int listen_fd)
 {
-    if (safe_mce_sys().actual_nginx_workers_num <= 0) {
-        return 0;
-    }
+    srdr_logdbg("worker: %d fd: %d", worker_id, listen_fd);
 
-    DO_GLOBAL_CTORS();
+    int ret = 0;
+    socket_fd_api *child_sock_fd_api = nullptr;
+    int parent_fd = listen_fd;
+    fd_collection *p_fd_collection = (fd_collection *)g_p_app->context;
+
+    /* Find information about parent socket
+     * Envoy:
+     * - g_p_fd_collection has all actual fds
+     * - worker 0 has socket object in g_p_fd_collection (parent fd)
+     * - Other workers should find parent fd to create child socket objects
+     *   basing on parent socket objects.
+     * Nginx:
+     * - should use fd_collection from parent process passed using g_p_app->context
+     */
+    if (g_p_app->type == APP_ENVOY) {
+        p_fd_collection = g_p_fd_collection;
+        if (worker_id > 0) {
+            parent_fd = -1;
+            for (const auto &itr1 : g_p_app->map_thread_id) {
+                if (itr1.second == 0) {
+                    for (const auto &itr2 : g_p_app->map_listen_fd) {
+                        if (itr2.second == itr1.first) {
+                            parent_fd = itr2.first;
+                            auto itr3 = g_p_app->map_dup_fd.find(itr2.first);
+                            if (itr3 != g_p_app->map_dup_fd.end()) {
+                                parent_fd = itr3->second;
+                            }
+                            break;
+                        }
+                    }
+                    break;
+                }
+            }
+            if (parent_fd < 0) {
+                return -1;
+            }
+        } else {
+            ret = -1;
+            child_sock_fd_api = p_fd_collection->get_sockfd(listen_fd);
+            if (child_sock_fd_api) {
+                ret = child_sock_fd_api->listen(child_sock_fd_api->m_back_log);
+            }
+            return ret;
+        }
+    }
 
-    if (!g_p_fd_collection_parent_process) {
-        return 0;
+    /* This part should be ignored by Envoy worker 0
+     * Envoy: parent_fd is fd of parent socket
+     * Nginx: parent_fd is equal to listen_fd
+     */
+    sockinfo *si;
+    socket_fd_api *parent_sock_fd_api = p_fd_collection->get_sockfd(parent_fd);
+    if (!parent_sock_fd_api || !(si = dynamic_cast<sockinfo *>(parent_sock_fd_api))) {
+        srdr_logerr("parent sockinfo is not found");
+        return -1;
     }
 
-    srdr_logdbg("g_worker_index: %d Size is: %d", g_worker_index,
-                g_p_fd_collection_parent_process->get_fd_map_size());
+    int block_type = si->is_blocking() ? 0 : SOCK_NONBLOCK;
+    sock_addr sa;
+    socklen_t sa_len = sa.get_socklen();
 
-    std::unordered_map<uint16_t, uint16_t> udp_sockets_per_port;
-    sockinfo *si;
-    for (int sock_fd = 0; sock_fd < g_p_fd_collection_size_parent_process; sock_fd++) {
```

Comment:
you removed that loop over all possible fds, but didn't implement it.
how does nginx works after that change?

Reply:
See special commit as https://github.com/Mellanox/libxlio/pull/25/commits/455c719f513f5c9ef077822e2a6e1d80a525149e
There is `map_listen_fd` for this